mlir/lib/Tools/PDLL/Parser/Parser.cpp

```diff
@@ -2245,6 +2255,114 @@ FailureOr<ast::Expr *> Parser::parseUnderscoreExpr() {
   return createInlineVariableExpr(type, name, nameLoc, constraints);
 }
 
+FailureOr<ast::Expr *> Parser::parseDictExpr() {
+  consumeToken(Token::l_brace);
+  SMRange loc = curToken.getLoc();
+
+  if (parserContext != ParserContext::Rewrite)
+    return emitError(
+        "Parsing of dictionary attributes as constraint not supported!");
+
+  auto dictAttrCall = createNativeCall(loc, "createDictionaryAttr", {});
+  if (failed(dictAttrCall))
+    return failure();
+
+  // Add each nested attribute to the dict
+  do {
+    FailureOr<ast::NamedAttributeDecl *> decl =
+        parseNamedAttributeDecl(llvm::None);
+    if (failed(decl))
+      return failure();
+
+    ast::NamedAttributeDecl *namedDecl = *decl;
+
+    std::string stringAttrValue =
+        "\"" + std::string((*namedDecl).getName().getName()) + "\"";
+    auto *stringAttr = ast::AttributeExpr::create(ctx, loc, stringAttrValue);
+
+    // Declare it as a variable
+    std::string anonName =
+        llvm::formatv("dict{0}", anonymousDeclNameCounter++).str();
+    FailureOr<ast::VariableDecl *> stringAttrDecl =
+        createVariableDecl(anonName, namedDecl->getLoc(), stringAttr, {});
+    if (failed(stringAttrDecl))
+      return failure();
+
+    // Get its reference
+    auto stringAttrRef = parseDeclRefExpr(
+        (*stringAttrDecl)->getName().getName(), namedDecl->getLoc());
+    if (failed(stringAttrRef))
+      return failure();
+
+    // Create addEntryToDictionaryAttr native call.
+    SmallVector<ast::Expr *> arrayAttrArgs{*dictAttrCall, *stringAttrRef,
+                                           namedDecl->getValue()};
+    auto entryToDictionaryCall =
+        createNativeCall(loc, "addEntryToDictionaryAttr", arrayAttrArgs);
+    if (failed(entryToDictionaryCall))
+      return failure();
+
+    // Uses the new array for the next element.
+    dictAttrCall = entryToDictionaryCall;
+  } while (consumeIf(Token::comma));
+  if (failed(parseToken(Token::r_brace,
+                        "expected `}` to close dictionary attribute")))
+    return failure();
+  return dictAttrCall;
+}
+
+FailureOr<ast::Expr *> Parser::parseArrayAttrExpr() {
+
+  // Advance to the next token without failing.
+  auto nextToken = [&](Token &curToken, int64_t offset) {
+    SMRange loc = curToken.getLoc();
+    SMRange dictArraysLoc(
+        loc.Start.getFromPointer(loc.Start.getPointer() + offset),
+        curToken.getEndLoc());
+    resetToken(dictArraysLoc);
+  };
+
+  SMRange loc = curToken.getLoc();
+
+  const char tokenAfterArray = *loc.End.getPointer();
+  if (tokenAfterArray != '[')
+    return emitError(curToken.getLoc(), "expected `[` after `array`.");
+
+  // Consume `array[` token by advancing 6 characters.
+  // Since the lexer misinterprets `[{` as a string_block, we can't consume the
+  // array token in the normal way. Instead, advance to the next token without
+  // looking at the new Token::Kind.
+  nextToken(curToken, 6);
```

Comment:
Here we are mixing stuff into the parser that I feel really belongs into the lexer. 
I would rather avoid blurring the clear separation (modulo backtracking when there are identifiers that have a keyword as prefix) of these.

Especially the `nextToken(curToken, 6);` is worrying to me. 
Usually you can insert stuff like newlines anywhere in PDLL to improve readability. With this approach the following is parsed correctly:
```PDLL
let newRoot = op<test.success>() { some_array = 
                                        array[attr<"\"test1\"">]} -> ();
```
but this is not anymore:
```PDLL
let newRoot = op<test.success>() { some_array = array
                                              [attr<"\"test1\"">]} -> ();
```
`error: expected [ after array.`

Would it work for this if we lex the keyword kw_array as `array[` instead of `array`?
If this workaround does not help I think it makes much more sense to extend the lexer in a sensible way

Reply:
As discussed, I will remove the keyword `array` and just have the `[` to express arrays.